### PR TITLE
[oauth2] Fix token refresh timeout cancellation with singleshot timer

### PR DESCRIPTION
**Pull request by dakcarto**
## Description
Fix token refresh timeout cancellation with singleshot timer.

- Previously, if there was an error that caused the `O2::refreshFinished` signal not to be fired, the interface would freeze due to `QEventLoop::ExcludeUserInputEvents` when running the loop. This ensures there is no input events ignored and that the loop is quit regardless of any refresh response issue.

Also, avoid O2 library error by checking for refresh token and URL first.

- This led to issues if the token refresh URL was optionally left blank; the refresh would still be attempted when it should be skipped and a new access token be requested instead.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/auth/oauth2/qgsauthoauth2method.cpp [hunk 1]
```diff
@@ -142,23 +143,50 @@ bool QgsAuthOAuth2Method::updateNetworkRequest( QNetworkRequest &request, const
...
+        // add singlshot timer to quit refresh after an alloted timeout
+        // this should keep the local event loop from blocking forever
+        QTimer r_timer( nullptr );
+        int r_reqtimeout = o2->oauth2config()->requestTimeout() * 1000;
+        r_timer.setInterval( r_reqtimeout );
+        r_timer.setSingleShot( true );
+        connect( &r_timer, &QTimer::timeout, &rloop, &QEventLoop::quit );
+        r_timer.start();
+
+        QgsTemporaryCursorOverride waitCursor( Qt::WaitCursor );
```
Comment (nyalldawson) [inline src/auth/oauth2/qgsauthoauth2method.cpp]: This looks dangerous. As far as I can see, there's nothing guaranteeing this is being run in the main thread -- so you're accessing gui methods from a background thread.

File: src/auth/oauth2/qgsauthoauth2method.cpp [hunk 2]
```diff
@@ -142,23 +143,50 @@ bool QgsAuthOAuth2Method::updateNetworkRequest( QNetworkRequest &request, const
...
-      // Asynchronously attempt the refresh
-      // TODO: This already has a timed reply setup in O2 base class (and in QgsNetworkAccessManager!)
-      //       May need to address this or app crashes will occur!
-      o2->refresh();
-
-      // block request update until asynchronous linking loop is quit
-      rloop.exec( QEventLoop::ExcludeUserInputEvents );
-
-      // refresh result should set o2 to (un)linked
+      if ( o2->refreshToken().isEmpty() || o2->refreshTokenUrl().isEmpty() )
```
Comment (nyalldawson) [inline src/auth/oauth2/qgsauthoauth2method.cpp]: My understanding is that it's NOT a requirement to have an explicit refresh url for o2.

File: src/auth/oauth2/qgsauthoauth2method.cpp (hunk 1 above)
Comment (dakcarto) [reply src/auth/oauth2/qgsauthoauth2method.cpp]: Ah, good point. Will fix.

File: src/auth/oauth2/qgsauthoauth2method.cpp (hunk 2 above)
Comment (dakcarto) [reply src/auth/oauth2/qgsauthoauth2method.cpp]: The O2 library checks for these as well. I am doing an up-front check here, so the library does not unnecessarily throw a network error during the aborted refresh operation:
https://github.com/qgis/QGIS/blob/master/external/o2/src/o2.cpp#L412-L421

File: src/auth/oauth2/qgsauthoauth2method.cpp (hunk 1 above)
Comment (dakcarto) [reply src/auth/oauth2/qgsauthoauth2method.cpp]: Removed cursor override.

**Comment by dakcarto** [comment]
Hmm. Travis CI has completed successfully over 4 hours ago, but the check still says Pending.

**Comment by dakcarto** [comment]
@nyalldawson, ok, CI seems to be reporting correctly now.

I removed the GUI call(s) from the auth method.

File: src/auth/oauth2/qgsauthoauth2method.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/auth/oauth2/qgsauthoauth2method.cpp]: Ah - i see what's happening. If an explicit refresh url isn't specified, o2 sets it to the token url. So all good here.

**Comment by nyalldawson** [comment]
@dakcarto looks good. 

I seem to get a LOT of odd timeouts here (not caused by this PR). I wonder if the o2 timeout mechanism is clashing with QgsNetworkAccessManager's inbuilt timeout handling...

**Comment by nyalldawson** [comment]
Actually - I think i can see the problem. The o2 handler is created on the first network request which needs it, which often happens in a background thread. This means the o2 handler can get orphaned when this thread exits, and it's event loop is no longer running, and all o2 requests go nowhere after that, because their queued connections are never processed.

We need a way to either:
1. ensure the handler is always created on the main thread
or
2. force it to be created in its own dedicated background thread which has an event loop that runs forever....

**Comment by dakcarto** [comment]
@nyalldawson wrote:
> I seem to get a LOT of odd timeouts here (not caused by this PR). I wonder if the o2 timeout mechanism is clashing with QgsNetworkAccessManager's inbuilt timeout handling...

Which QGIS distro/platform (if not building yourself) are you testing here? I noticed there is something **very** different about the Kyngchaos distro (LTR or latest) that throws a lot of OAuth2 timeouts. The same OAuth2 connections in OSGeo4W or Lutra QGIS.app on macOS, or builds upon `conda` dependencies, do not exhibit the same issue.

I was hoping your latest [threading commit](https://github.com/qgis/QGIS/commit/9cde65457b3717a74e32277c843a2945eb4f1221#diff-bccd6a598a93e77a6afef8e5679ae512) would solve this issues with Kyngchaos distro, but have not tested yet with 3.4.8 release.

**Pull request merged**